### PR TITLE
[go-experimental][go] Allows overriding base URL

### LINKED ISSUE #5818: [go-experimental] Allow overriding server URL
### Is your feature request related to a problem? Please describe.

Since #4755 the go-experimental generator supports selecting servers, but I can't see how to override server URLs with an arbitrary basepaths. That would be useful, for example, when testing a client generated with production spec against an in-development version of a server.

## Describe the solution you'd like

Having a context option for setting an arbitrary basepath would be nice. This maintains the spirit of #4755 of configuring the client via context options.

It would be configured like so:

```golang
ctx := context.WithValue(context.Background(), sw.ContextBasepath, "http://localhost:8000/in-development")
```

### PR DESCRIPTION
**Pull request by thiagoarrais**
Fixes #5818 

cc @antihax @bvwells @grokify @kemokemo @bkabrda

### CODE REVIEW

**Comment by thiagoarrais** [comment]
Just rebased this to latest master.

@bkabrda and @jirikuncar: would love to hear your opinion here.

File: samples/client/petstore/go-experimental/go-petstore/configuration.go [hunk 1]
```diff
@@ -45,6 +45,9 @@ var (
 	// ContextServerIndex uses a server configuration from the index.
 	ContextServerIndex = contextKey("serverIndex")
 
+	// ContextBasepath overrides the API base path
+	ContextBasepath = contextKey("basepath")
```
Comment (jirikuncar) [inline samples/client/petstore/go-experimental/go-petstore/configuration.go:49]:
```suggestion
	ContextBasePath = contextKey("basePath")
```

File: modules/openapi-generator/src/main/resources/go-experimental/configuration.mustache [hunk 2]
```diff
@@ -275,6 +278,14 @@ func (c *Configuration) ServerURLWithContext(ctx context.Context, endpoint strin
 		return sc.URL(0, nil)
 	}
 
+	v := ctx.Value(ContextBasepath)
```
Comment (jirikuncar) [inline modules/openapi-generator/src/main/resources/go-experimental/configuration.mustache:281]:
```suggestion
	v := ctx.Value(ContextBasePath)
```

File: modules/openapi-generator/src/main/resources/go-experimental/configuration.mustache [hunk 3]
```diff
@@ -275,6 +278,14 @@ func (c *Configuration) ServerURLWithContext(ctx context.Context, endpoint strin
...
+	if v != nil {
+		if basepath, ok := v.(string); ok {
+			return basepath, nil
+		}
+		return "", reportError("ctx value of ContextBasepath has Invalid type %T should be string", v)
```
Comment (jirikuncar) [inline modules/openapi-generator/src/main/resources/go-experimental/configuration.mustache:286]:
```suggestion
		return "", reportError("ctx value of ContextBasePath has Invalid type %T should be string", v)
```

File: modules/openapi-generator/src/main/resources/go-experimental/configuration.mustache [hunk 4]
```diff
@@ -37,6 +37,9 @@ var (
 	// ContextServerIndex uses a server configuration from the index.
 	ContextServerIndex = contextKey("serverIndex")
 
+	// ContextBasepath overrides the API base path
+	ContextBasepath = contextKey("basepath")
```
Comment (jirikuncar) [inline modules/openapi-generator/src/main/resources/go-experimental/configuration.mustache:41]:
```suggestion
	ContextBasePath = contextKey("basePath")
```

**Comment by thiagoarrais** [comment]
good idea, @jirikuncar. I've added a warning. But I'm not really sure I've done it "correctly". I couldn't find any other instance of warnings in go-experimental.

**Comment by thiagoarrais** [comment]
Weird: build is green in Travis CI, but the PR doesn't get updated in Github

**Comment by thiagoarrais** [comment]
Rebased to current master. Anything else needed here?

**Comment by thiagoarrais** [comment]
This PR may not be needed. I hadn't noticed there is a `Servers` component in `Configuration`.

The base URL may be overriden (in master itself) like so:

```go
	apiConfig := openapi.Configuration{
		Servers: openapi.ServerConfigurations{
			{
				URL: basepath,
			},
		},
	}

	openAPIClient := openapi.NewAPIClient(&apiConfig)
```
I haven't tested this extensively, but it seems to work well for the specific API I'm interested in. I will leave this PR open for a week in order to gather some feedback about this alternative.

**Pull request closed without merging**